### PR TITLE
FFT Convolution

### PR DESCRIPTION
**Pull request by peterroelants**
I took  a stab at implementing FFT based convolutions as an option in `jax.scipy.signal.convolve`. The implementation is largely based on scipy's [`fftconvolve`](https://github.com/scipy/scipy/blob/v1.6.2/scipy/signal/signaltools.py#L551-L663).

## Contributions
- Allow passing not None `shape` parameter to `fft._fft_core`, which is needed to get the FFT-based convolutions to work.
- `jax.scipy.signal.fftconvolve` operator similar to `scipy.signal.fftconvolve`
- Allow `'direct'` and `'fft`' as methods in `jax.scipy.signal.convolve`
- Tests for the changes

I didn't implement the logic for `'auto`' to decide between `'direct'` or `'fft'` yet. `'auto'` falls back to the `'direct'` method while giving a warning for now.

## Example
Please see [this notebook](https://nbviewer.jupyter.org/urls/gist.githubusercontent.com/peterroelants/e7e9cdde0006b9809cf203cd41419e35/raw/8b62593b2495a9213f9deb092aef01b53f13a6d1/jax_fftconvolve.ipynb) for an example. Note that for an array of size 10k the `'fft'` based convolution is on the order of 40x faster compared to the `'direct'` convolution.

## Feedback Wanted
I separated out the `jax.numpy.fft` changes from the `jax.scipy.signal` contributions. I hope this is ok, if not I can split this up in two different PRs, or merge the commits.

I'd love to get some feedback on this and hope to eventually merge this contribution into Jax.

### CODE REVIEW

**Comment by peterroelants** [comment]
@googlebot I fixed it

**Comment by jakevdp** [comment]
Hi @peterroelants – thanks so much for looking into this! It looks like a solid implementation of FFT convolve at the `jax.scipy` level.

That said, there is a difference between JAX and scipy that is relevant here: in JAX, we generally rely on the XLA compiler rather than the user to choose the most efficient computational path for any operation. So in a perfect world, XLA would be able to choose whether a given convolution would be more efficiently computed via FFT or via direct convolution. XLA does not yet implement this code path for most backends (the exception, I believe, is the GPU backend where CUDNN may use FFT for convolutions where it makes sense).

Since that's the future goal, we should probably future-proof this code and avoid putting this kind of logic at such a high level within JAX.

Where this approach might make sense is in a CPU translation rule for `lax.conv_general_dilated`. The idea there would be that, on CPU, we could lower convolution operations to an FFT implementation in cases where it makes sense. That would lead to two advantages:

1. It would improve runtimes for *all* relevant convolutions in JAX, not just those used via `scipy.signal`
2. For users, it would avoid the cognitive overhead of having to choose which FFT arguments to pass
2. In an eventual future where FFT convolutions are implemented in XLA itself, we could straightforwardly take advantage of that by adjusting the translation rule, rather than making changes to the user-facing API.

You can see an example of this kind of approach in many of the `lax` linear algebra routines, where CPU translation rules are used to implement various operations in terms of LAPACK calls; for example here: https://github.com/google/jax/blob/3c1ee0644b580b1c7cf7fd32043916da7b487390/jax/_src/lax/linalg.py#L339-L340

What do you think?

**Comment by shoyer** [comment]
@jakevdp I agree with the long-term vision of such an approach, but would we also add a new parameter to `conv_general_dilated` for controlling the convolution implementation? This would be needed for `scipy.signal.convolve`. I don't know if we can predict ahead of time which method is best.

**Comment by peterroelants** [comment]
Hi @jakevdp , thank you for the quick feedback, it's appriciated.

I'm trying to wrap my head around your comment, so let me try to summarize to make sure I understand you:
1. Instead of having the `method` parameter being used by the user you would like it fixed on `'auto'` and make sure the underlying implementation deals with with deciding to use the FFT optimization or not?
2. It seems that the `method` parameter in the `convolve` and `correlate` methods is only there because of compatibility with Scipy's API?
3. Making the decision to use the FFT optimization or not best happens in a translation rule for [`jax.lax.conv_general_dilated`](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.conv_general_dilated.html)

What I'm still uncertain of:
4. Are you suggesting to write a FFT version of `conv_general_dilated` with all its cases? Or would it be possible to start with a subset (e.g. only real-valued, non-dilated convolutions)?
5. I see that there is already a [translation rule for `conv_general_dilated`](https://github.com/google/jax/blob/5a336c6e246fb51e312cdda7a0a1432fffc689d9/jax/_src/lax/lax.py#L2922). Would the logic to decided between FFT convolution or direct convolution be best placed in this method?
6. How to make the decision between FFT optimization or not?

I'm quite novice to Lax. I might try following your suggestions the next weeks in between work to see what I can do. However, if I feel I'm out of my depth and it might take me too long to come up with a clean solution I might leave this PR as-it-is and revert to a custom implementation for my personal use case.

**Comment by jakevdp** [comment]
> Are you suggesting to write a FFT version of conv_general_dilated with all its cases? Or would it be possible to start with a subset (e.g. only real-valued, non-dilated convolutions)?

No, since this is purely an optimization we could limit it to a subset of the cases handled by the general operator.

> I see that there is already a translation rule for conv_general_dilated. Would the logic to decided between FFT convolution or direct convolution be best placed in this method?

Yes, that would be the place for it!

> How to make the decision between FFT optimization or not?

That's a good question – I'm not sure! @shoyer makes a good point about maybe adding an argument to `conv_general_dilated` to let the caller choose the method.

File: jax/_src/numpy/fft.py [hunk 1]
```diff
@@ -62,6 +60,12 @@ def _fft_core(func_name, fft_type, a, s, axes, norm):
         s += [max(0, 2 * (a.shape[axes[-1]] - 1))]
     else:
       s = [a.shape[axis] for axis in axes]
+  else:
+    if fft_type == xla_client.FftType.IRFFT:
+      _s = tuple(list(s)[:-1] + [max(0, s[-1] // 2 + 1)])
+      a = _fix_shape(a, _s, axes)
+    else:
+      a = _fix_shape(a, s, axes)
```
Comment (shoyer) [inline jax/_src/numpy/fft.py:68]: If you get the chance, maybe take a look at https://github.com/google/jax/pull/6266, which is also trying to add support for setting `s` in FFT?

Ideally we would merge this change as a separate PR, first.

**Comment by shoyer** [comment]
> 6. How to make the decision between FFT optimization or not?

I think it would be best to start with keeping the default (at least in `conv_general_dilated`) as `method='direct'`, with the option to manually switch to `method='fft'`. Once we're sure that works well, we can look into heuristics for choosing the optimal approach. This will likely be more complex than SciPy, because it may depend on both the argument shapes and the specific computing platform (CPU vs GPU vs TPU).

One thing I"ll mention that could make this easier to write is the `xla.lower_fun_initial_style()` helper function, which lets you write translation rules in terms of `jax.lax` rather than the lower-level XLA API (`xops`).
https://github.com/google/jax/blob/634397dc5925270a543fcbbd8b5f9ef5e2f3e8a4/jax/_src/lax/control_flow.py#L2354-2355

File: jax/_src/numpy/fft.py (hunk 1 above)
Comment (peterroelants) [reply jax/_src/numpy/fft.py:68]: Yes, https://github.com/google/jax/pull/6266 seems to do largely the same as commit https://github.com/google/jax/pull/6343/commits/9291632ea331a4b149ebd6e798505b4c50b64a67 from this PR.

Differences from quickly reading through the code:
-  Explicit use of padding (theirs), vs copying in an array of zeros (mine)
- Only pad when needed (mine), vs pad when a change is needed (theirs). (Maybe `lax.pad` can actually decide not to do any padding operation when it notices this is not needed, I'm not familiar with these optimizations).
- Separate tests with and without shape (mine).
- I think their tests don't test combinations of both crop and pad (From quickly reading through the code, I might be wrong.)

In general I think they are roughly the same, and I can update this PR once https://github.com/google/jax/pull/6266 has been merged.

**Comment by PhilipVinc** [comment]
bump! is there any update for this?

**Comment by peterroelants** [comment]
> bump! is there any update for this?

I, personally, have not been able to rewrite this PR yet.

**Comment by martin-millon** [comment]
Hi, is there any update on this issue? Is there any chance that this PR will be merged?

**Comment by hawkinsp** [comment]
Closing as stale.
